design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
+
+In headless, the session is opened when starting the ksqlDB server with a SQL script. The variables stay alive until the server is
+killed.
+
+Variables are not persisted anywhere. They are living in memory as long as the session is not terminated.
+
+#### RUN SCRIPT command
+
+If a `RUN SCRIPT` command is executed in a CLI session, then any variable defined there will be local to the execution context of the
+script. The variables will override any variable previously defined in the session. When the script is completed, all local variables
+defined will be destroyed. This ensures users do not override their session variables by accident when running `RUN SCRIPT` commands.
+
+#### The SOURCE command
+
+To allow users load variables and keep their scope in the session, then the new `SOURCE` command will be added. This will override
+any previously defined variables, and will be destroyed until the user leaves the session.
+
+Syntax:
+```
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
+executed, thus to avoid limiting users from using DDL/DML commands while defining variables at the same time from one single script.
+
+`SOURCE` will be used only on the CLI and headless mode. REST API will not accept it because files may be out of the scope of the server filesystem.
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `${}` characters (i.e. `${replicas}`).
+
+Noe: We need to make changes in QTT tests to use `${}` references instead of `{}` for consistency.
+
+Example:
+```
+ksql> DEFINE format = 'AVRO';
+ksql> DEFINE replicas = '3';
+ksql> CREATE STREAM stream1 (id INT) WITH (kafka_topic='stream1', value_format='${format}', replicas=${replicas});
+```
+
+Substitution will not attempt to add single-quotes to the values. Users will need to know the data type to use when using a variable. The ksqlDB
+server will be in charge of parsing and failing if the type is not allowed.
+
+Note: Variables are case-insensitive. A reference to `${replicas}` is the same as `${REPLICAS}`.
+
+### 4. Context for substitution variables
+
+Variable substitution will be allowed in specific SQL statements. They can be used to replace text and non-text literals, and identifiers such as
+column names and stream/table names. Variables cannot be used as reserved keywords.
```

Comment:
Is it normal for variables to not be allowed as keywords? I don't mind but if we go with this we need to make sure we have good error messages since a natural extension (unless it's just my naivete speaking?) from the example below:
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
is to wonder about
```
ksql> CREATE STREAM ${streamName} (${colName1} ${colType1}, ${colName2} ${colType2}) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
which won't be accepted based on this proposal. Hopefully the error message makes it very clear why.

Reply:
That really depends. Some big-data DBs allow any substitution, but other DBs (mysql, oracle, ...) do not allow replacing keywords. They have a different way to do that, though. 

Here's the discussion with @colinhicks about it.
https://github.com/confluentinc/ksql/pull/6259#discussion_r491181382

Reply:
I think given the goal (running these scripts in different environments) it's probably OK to not allow the types

Reply:
Agreed it's fine to not allow keyword substitutions -- as long as we have good error messages. The error messages on syntax errors we have today are pretty abysmal... :)